### ISSUE TITLE
Git+ssh auth doesn't work correctly in bitnami/phabricator:2019

### ISSUE THREAD

**Issue by emempi** [bitnami/bitnami-docker-phabricator#128]
**Description**

sshd setup for git in the phabricator image is not correct; preventing from even simple use-cases to work. Below I provide workarounds for these, but not knowing the intended design, the real fix might need be different.

**Steps to reproduce the issue:**

1. Install bitnami phabricator image using instructions at https://hub.docker.com/r/bitnami/phabricator/
2. Log in as `user` and upload your public ssh key
3. Try cloning the test repo

**Describe the results you received:**
```
$ git clone ssh://git@127.0.0.1:2222/source/test.git
Cloning into 'test'...
git@127.0.0.1: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Describe the results you expected:**

Correctly cloned repository.

**Additional information you deem important (e.g. issue happens only occasionally):**

This has been caused by a multitude of issues listed below.

### Locked `git` account

The `git` account in the docker image is locked by default. Sshd prevents any logins to the github accounts:

```
debug3: account expiration disabled
User git not allowed because account is locked
debug3: mm_answer_pwnamallow: sending MONITOR_ANS_PWNAM: 0
debug3: mm_request_send entering: type 9
debug2: monitor_read: 8 used once, disabling now
input_userauth_request: invalid user git [preauth]
debug3: mm_audit_event entering [preauth]
debug3: mm_request_send entering: type 112 [preauth]
debug3: mm_inform_authserv entering [preauth]
debug3: mm_request_send entering: type 4 [preauth]
debug2: input_userauth_request: try method none [preauth]
debug3: userauth_finish: failure partial=0 next methods="publickey" [preauth]
```

Workaround:
Manually editing `/etc/shadow` and replacing:
```
git:!:18149:1:90:7:::
```

with

```
git:*:18149:1:90:7:::
```

### The `git` user doesn't have correctly configured environment

After implementing the previous workaround, the private key authentication fails with:

```
debug3: subprocess: AuthorizedKeysCommand command "/opt/bitnami/phabricator/resources/sshd/phabricator-ssh-hook.sh git" running as git
debug1: temporarily_use_uid: 1002/1002 (e=0/0)
debug1: restore_uid: 0/0
debug3: subprocess: AuthorizedKeysCommand pid 836
debug1: temporarily_use_uid: 1002/1002 (e=0/0)
debug2: key not found
AuthorizedKeysCommand /opt/bitnami/phabricator/resources/sshd/phabricator-ssh-hook.sh git failed, status 127
```

Invoking the AuthorizedKeysCommand directly, I'm getting:

```
# su - git /opt/bitnami/phabricator/resources/sshd/phabricator-ssh-hook.sh git
No directory, logging in with HOME=/
/usr/bin/env: 'php': No such file or directory
```
Workaround:
```
# ln -s /opt/bitnami/php/bin/php /usr/local/bin
```

### Incorrectly configured home directory for sshd:
```
$ git clone ssh://git@127.0.0.1:2222/source/test.git
Cloning into 'test'...
Environment:
  USER=git
  LOGNAME=git
  HOME=/home/phabricator
  PATH=/usr/local/bin:/usr/bin:/bin:/usr/games
  MAIL=/var/mail/git
  SHELL=/bin/sh
  SSH_CLIENT=172.18.0.1 45598 2222
  SSH_CONNECTION=172.18.0.1 45598 172.18.0.3 2222
  SSH_ORIGINAL_COMMAND=git-upload-pack '/source/test.git'
Could not chdir to home directory /home/phabricator: No such file or directory
warning: You appear to have cloned an empty repository.
```

This didn't prevent checkout of the repository (that already succeeded when applying the second workaround), but still issued a warning.

Workaround:

Manually editing `/etc/passwd` and changing `/home/phabricator` to /opt/phabricator` (I guess any existing directory with sufficient permissions would work).

**Version**

- Output of `docker version`:

```
Docker version 18.09.1, build 4c52b90
```

- Output of `docker info`:

```
Containers: 139
 Running: 2
 Paused: 0
 Stopped: 137
Images: 570
Server Version: 18.09.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 1.0.0~rc6+dfsg1-3
init version: v0.18.0 (expected: fec3683b971d9c3ef73f284f176672c44b448662)
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 5.2.0-2-amd64
Operating System: Debian GNU/Linux bullseye/sid
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.49GiB
Name: milan-plzik
ID: UIQV:IADD:77G3:KDLO:PAMT:2SNF:F4RB:ESLV:IRQO:ZORL:N6AM:HSKP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.21.0, build unknown
docker-py version: 3.4.1
CPython version: 3.7.4+
OpenSSL version: OpenSSL 1.1.1c  28 May 2019
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

Debian unstable, Docker.

**Comment by rafariossaa** [comment]
Hi,
  I am creating a internal task to fix this, we will come back to you as soon this is fixed or we have news.

**Comment by emempi** [comment]
Cool, thanks a lot! :)

**Comment by altendky** [comment]
Is there any way this can be opened up a bit more?  It's been over a year since I had something working but because of the secrets multiple people have had to rediscover various issues and workarounds instead of just fixing the core issue.

**Comment by altendky** [comment]
Or maybe there's a paid support option where we could get this feature fixed promptly?

**Comment by juan131** [comment]
Hi @altendky 

> Or maybe there's a paid support option where we could get this feature fixed promptly?

I'm afraid we do not provide any kind of personal/paid support for our community catalog.

We just released a new revision of Phabricator (**2019.38.0-r3**) that addresses your suggestions. could you please give it a try?

**Comment by altendky** [comment]
@juan131 we were able to get up and running with those changes.  Thanks for the work on this.  Hopefully @emempi can confirm as well.

**Comment by rafariossaa** [comment]
H @emempi ,
  Could you confirm that ?

**Comment by emempi** [comment]
I just tested the new image using the most recent `docker-compose.yml`, with added port mapping 2222:22 and env `      - PHABRICATOR_ENABLE_GIT_SSH_REPOSITORY=yes`; unfortunately, the sshd didn't seem to start at all. I don't have time to investigate much more now -- is this an expected behavior? (note: I did not do absolutely clean installation -- there already was mysql and phabricator data volume).

**Comment by juan131** [comment]
Hi @emempi 

> (note: I did not do absolutely clean installation -- there already was mysql and phabricator data volume).

That's weird. I mean, the current initialization logic configures SSH daemon to support hosted GIT repositories both when persisted data is detected and when it's not.

Did you update the tag for Phabricator or are you using "latest"?

**Comment by emempi** [comment]
Hello, I used 'bitnami/phabricator:2019', which was at that point corresponding to image `232065e07df3`.

**Comment by juan131** [comment]
Hi @emempi 

> I used `bitnami/phabricator:2019`

That's a rolling tag and I don't recommend you to use it if you're running this Phabricator container on a production environment. You can find more information in the link below:

- https://docs.bitnami.com/containers/how-to/understand-rolling-tags-containers/

Could you please try using the specific tag revision?

**Comment by emempi** [comment]
Looks good; I was able to easily create git repo, clone it and push to it.

As for the floating tag, I think the problem was elsewhere in the end. I made sure to use --no-cache when fetching `2019`, so I should have gotten the freshest build if the tags are being updated correctly. Unfortunately, there was a typo in my hastily-created config that prevented stuff from working correctly. Once this got fixed, everything started working. Thanks for fixing the issue :)
